chainer/graph_optimizations/static_graph.py

```diff
@@ -0,0 +1,1464 @@
+                self.schedules[key_str] = [sched]
+                self.in_use_count[key_str] = 1
+
+        return sched
+
+    def end_forward(self):
+        """Make in-use schedules available for use in next iteration.
+
+        Set the in-use status of all schedules to "not in use" so that
+        they can be reused in the next iteration.
+
+        In the case that test mode is active
+        (`chainer.config.train` is `False`) and the static chain corresponding
+        to this manager was not called more than once in any iteration during
+        training mode, then this method will be called automatically.
+
+        """
+        if not self.forward_over:
+            for key in self.in_use_count:
+                self.in_use_count[key] = 0
+            self.forward_over = True
+
+            if self.train_count > self.max_in_use_train:
+                self.max_in_use_train = self.train_count
+                if self.verbosity_level >= 2:
+                    print("Maximum in-use schedules per training iteration: ",
+                          self.max_in_use_train)
+            self.train_count = 0
+
+    def __repr__(self):
+        out = "ScheduleManager:\n"
+        for key_str in self.schedules:
+            out += "key string: " + key_str
+            sched_list = self.schedules[key_str]
+            out += " -> schedule list of length: " + \
+                   str(len(sched_list)) + '\n'
+            for sched in sched_list:
+                out += str(sched)
+        return out
+
+
+def static_graph(*args, **kwargs):
+    """Decorator to mark a Chain's ``__call__()`` as a static sub-graph.
+
+    This decorator marks the define-by-run code inside the `__call__()`
+    method of a Chain instance as corresponding to a static computation
+    graph or sub-graph. Such a chain will be referred to as a 'static chain'.
+    This allows various "static graph" optimizations to be performed, which
+    can result in significant speedups for some models.
+
+    When this decorator is used, the chain's define-by-run code executes
+    during the first iteration as usual. However, while the define-by-run
+    code is executing, a trace is also performed to incrementally create a
+    corresponding static schedule. This static schedule will only contain
+    the subset of the computations inside the define-by-run code that actually
+    needs to run every iteration. Specifically, this will contain the code
+    inside any functions called that were annotated with the `@static_code`
+    decorator, which will include all Chainer built-in functions, as well as
+    any user-defined functions that use `@static_code`. Then, starting
+    from the second iteration, when the static chain is called, its
+    static schedule code will be executed instead of its define-by-run code.
+
+    However, the user must also be careful of the following:
+    - The user is responsible for applying this decorator correctly. The
```

Comment:
These bullet points are not shown appropriately when I read this doc in HTML.

Reply:
thanks. I'll check.